src/main/java/pl/edu/agh/firecell/engine/algorithm/FirePropagator.java

```diff
@@ -0,0 +1,103 @@
+        };
+    }
+
+    public static Stream<Vector3i> doesNeighbourOnFire(State oldState, Vector3i cellIndex){
+        return Stream
+                .concat(NeighbourUtils.neighboursStream(cellIndex, NeighbourUtils.Axis.X), NeighbourUtils.neighboursStream(cellIndex, NeighbourUtils.Axis.Z))
+                .filter(neighbourIndex -> oldState.hasCell(neighbourIndex) &&
+                        isOnFire(oldState.getCell(neighbourIndex)) &&
+                        oldState.getCell(neighbourIndex).remainingFirePillar() > 0 &&
+                        !possibleToGoUp(oldState, neighbourIndex));
+    }
+
+    public int computeFirePillar(State oldState, Cell oldCell, Vector3i cellIndex, int currentFirePillar){
+        if (!oldCell.material().equals(AIR))
+            return currentFirePillar;
+
+        // from under
+        int downFirePillar = 0;
+        if (oldState.hasCell(NeighbourUtils.down(cellIndex)) &&
+                isOnFire(oldState.getCell(NeighbourUtils.down(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.down(cellIndex)).remainingFirePillar() - 1 > 0) {
+            downFirePillar = oldState.getCell(NeighbourUtils.down(cellIndex)).remainingFirePillar() - 1;
+        }
+
+        // from neighbour
+        int neighbourFirePillar = 0;
+        try {
+            int tmp = Collections.max(doesNeighbourOnFire(oldState, cellIndex)
+                    .filter(neighbourIndex -> !possibleToGoUp(oldState, neighbourIndex) &&
+                            oldState.getCell(neighbourIndex).remainingFirePillar() - 1 > 0)
+                    .map(neighbourIndex -> oldState.getCell(neighbourIndex).remainingFirePillar())
+                    .toList());
+            if(tmp>currentFirePillar){
+                neighbourFirePillar = tmp-1;
+            }
+        } catch (NoSuchElementException ignored) {}
+
+        return Math.max(downFirePillar, neighbourFirePillar);
+    }
+
+    private int newAirBurningTime(State oldState, Vector3i cellIndex) {
+        int newBurningTime;
+        newBurningTime = 0;
+
+        if (oldState.hasCell(NeighbourUtils.down(cellIndex)) &&
+                isOnFire(oldState.getCell(NeighbourUtils.down(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.down(cellIndex)).remainingFirePillar() - 1 > 0) {
+            newBurningTime++;
+        }
+
+        if (!oldState.hasCell(NeighbourUtils.down(cellIndex)) ||
+                !isOnFire(oldState.getCell(NeighbourUtils.down(cellIndex))))
+        {
+            // cell under is not on fire or not present
+            if(doesNeighbourOnFire(oldState, cellIndex)
+                    .anyMatch(neighbourIndex -> !possibleToGoUp(oldState, neighbourIndex) &&
+                            oldState.getCell(neighbourIndex).remainingFirePillar() - 1 > 0))
+                newBurningTime = 1;
+        }
+        return newBurningTime;
+    }
+
+    private int newWoodBurningTime(Cell oldCell, double newTemperature, int currenBurningTime) {
```

Comment:
Don't reassign arguments. It does not change the result in this case (int values are passed by value) but is a really bad practise.